internal/webhooks/machinedrainrules.go

```diff
@@ -0,0 +1,135 @@
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&clusterv1.MachineDrainRule{}).
+		WithValidator(webhook).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1beta1-machinedrainrule,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=machinedrainrules,versions=v1beta1,name=validation.machinedrainrule.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+// MachineDrainRule implements a validation webhook for MachineDrainRule.
+type MachineDrainRule struct{}
+
+var _ webhook.CustomValidator = &MachineDrainRule{}
+
+// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type.
+func (webhook *MachineDrainRule) ValidateCreate(_ context.Context, obj runtime.Object) (admission.Warnings, error) {
+	mdr, ok := obj.(*clusterv1.MachineDrainRule)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a MachineDrainRule but got a %T", obj))
+	}
+
+	return nil, webhook.validate(mdr)
+}
+
+// ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type.
+func (webhook *MachineDrainRule) ValidateUpdate(_ context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+	newMDR, ok := newObj.(*clusterv1.MachineDrainRule)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a MachineDrainRule but got a %T", newObj))
+	}
+
+	return nil, webhook.validate(newMDR)
+}
+
+// ValidateDelete implements webhook.CustomValidator so a webhook will be registered for the type.
+func (webhook *MachineDrainRule) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (webhook *MachineDrainRule) validate(newMDR *clusterv1.MachineDrainRule) error {
+	var allErrs field.ErrorList
+
+	if newMDR.Spec.Drain.Behavior == clusterv1.MachineDrainRuleDrainBehaviorSkip {
+		if newMDR.Spec.Drain.Order != nil {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec", "drain", "order"),
+					*newMDR.Spec.Drain.Order,
+					fmt.Sprintf("order must not be set if drain behavior is %q", clusterv1.MachineDrainRuleDrainBehaviorSkip)),
+			)
+		}
+	}
+
+	allErrs = append(allErrs, ValidateMachineDrainRulesSelectors(newMDR)...)
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("MachineDrainRule").GroupKind(), newMDR.Name, allErrs)
+}
+
+// ValidateMachineDrainRulesSelectors validate the selectors of a MachineDrainRule.
+// Note: This func is exported so it can be also used to validate selectors in the Machine controller.
+func ValidateMachineDrainRulesSelectors(machineDrainRule *clusterv1.MachineDrainRule) field.ErrorList {
+	var allErrs field.ErrorList
+
+	for i, machineSelector := range machineDrainRule.Spec.Machines {
+		if machineSelector.Selector != nil {
+			if _, err := metav1.LabelSelectorAsSelector(machineSelector.Selector); err != nil {
```

Comment:
Sadly this cannot be validated using CEL yet, but it is on the wishlist. If it's implemented in the future, then as far as I can tell, this entire webhook could be converted to CEL

Reply:
Agree. We'll still have to reimplement the same code for our controller, but we could get rid of the webhook